## **NEST & NOOK RETAIL**

<img src="nest_and_nook_image.jpeg" alt="nest_nook" width="300" height="10">

## **BUSINESS OVERVIEW**
**Nest & Nook**, a leading online retail business, presents a comprehensive dataset on shopping trends spanning from December 2010 to December 2011. This extensive dataset captures detailed shopping information from across a global network of 38 countries. 

## **PROJECT OBJECTIVE**

The objective of this project is to perform a customer segmentation and profiling analysis using the RFM (Recency, Frequency, Monetary) model. This analysis will help us identify distinct customer segments, profile their shopping behaviors, and pinpoint high-value customers. The insights gained will optimize marketing strategies, enhance customer experiences, and support data-driven decisions to drive business growth in **Nest & Nook.**.

- Data covers orders between **December 1st, 2010** and **December 9th, 2011.**

## **SALES TABLE DESCRIPTION**

<img src="nest_and_nook_data_desc.jpeg" alt="sales_table" width="700" height="50">

## **CREATING SALES TABLE**

### **Creating Schema**

In [11]:
%%sql
-- CREATING ORDERS TABLE
CREATE TABLE orders (
    invoice_no VARCHAR(10),
    invoice_date DATE,
    invoice_time TIME,
    stock_code VARCHAR(20),
    description VARCHAR(100),
    quantity INT,
    unit_price NUMERIC,
    total_sales NUMERIC,
    customer_id INT
);

-- CREATING CUSTOMERS TABLE
CREATE TABLE customer (
    customer_id INT PRIMARY KEY,
    country VARCHAR(50),
    gender VARCHAR(10) CHECK (gender IN ('Male', 'Female')),
    age INT CHECK (age > 0)
);

 * postgresql://postgres:***@localhost:5432/rfm_nest_and_nook
(psycopg2.errors.DuplicateTable) relation "orders" already exists

[SQL: -- CREATING ORDERS TABLE
CREATE TABLE orders (
    invoice_no VARCHAR(10),
    invoice_date DATE,
    invoice_time TIME,
    stock_code VARCHAR(20),
    description VARCHAR(100),
    quantity INT,
    unit_price NUMERIC,
    total_sales NUMERIC,
    customer_id INT
);]
(Background on this error at: https://sqlalche.me/e/20/f405)


### **Importing CSV File Into Schema**

In [9]:
%%sql
-- import orders csv into orders table schema
COPY orders (invoice_no, invoice_date, invoice_time, stock_code, description, quantity, unit_price, total_sales, customer_id)
FROM 'C:\Program Files\PostgreSQL\16\scripts\Dataset\nest_and_nook\nest_and_nook_online_retail_2.csv'
DELIMITER ','
CSV HEADER;

-- import customers records into customers table schema
COPY customer (customer_id, country, gender, age)
FROM 'C:\Program Files\PostgreSQL\16\scripts\Dataset\nest_and_nook\nest_and_nook_customer.csv'
DELIMITER ','
CSV HEADER;

 * postgresql://postgres:***@localhost:5432/rfm_nest_and_nook
541909 rows affected.


[]

## **RFM ANALYSIS**

**RFM analysis** is a way to understand and categorize customers based on their past buying habits. It uses three main factors:
- **Recency (R)**: How recently a customer made a purchase. More recent purchases get higher scores.
- **Frequency (F)**: How often a customer makes purchases. Frequent buyers get higher scores.
- **Monetary Value (M)**: How much money a customer spends. Bigger spenders get higher scores.

## **Recency**
- This represents the duration of time (in days) that has passed since a customer’s last purchase. This is calculated by subtracting each customer's last purchase date from the last order date in the data (that is December 9th, 2011).

In [80]:
%%sql
-- CREATING A VIEW TO STORE EACH CUSTOMER AND THEIR RECENCY SCORE
CREATE VIEW r_score_view AS (
-- selecting the latest records for each customer 
    WITH date_rank AS (
    SELECT
        customer_id,
        invoice_date,
        ROW_NUMBER() OVER(PARTITION BY customer_id ORDER BY invoice_date DESC) AS rank
    FROM orders
),
-- subtracting max date from the latest date each customer purchased
recency_1 AS (
    SELECT
        customer_id,
        '2011-12-09' - invoice_date AS recency
FROM date_rank
WHERE rank = 1
),
-- grouping the recency into 4 quartiles
recency_2 AS (
    SELECT
        customer_id,
        recency,
        NTILE(4) OVER(ORDER BY recency DESC) AS r_score
    FROM recency_1
)
SELECT *
FROM recency_2
ORDER BY customer_id
)

 * postgresql://postgres:***@localhost:5432/rfm_nest_and_nook
Done.


[]

In [81]:
%%sql
-- previewing the r_score_view
SELECT *
FROM r_score_view
LIMIT 5

 * postgresql://postgres:***@localhost:5432/rfm_nest_and_nook
5 rows affected.


[(12346, 325, 1),
 (12347, 2, 4),
 (12348, 75, 2),
 (12349, 18, 3),
 (12350, 310, 1)]

## **Frequency**
- This is calculated by counting the number of times a customer bought a product.

In [94]:
%%sql
-- CREATING A VIEW TO STORE EACH CUSTOMER AND THEIR FREQUENCY SCORE
CREATE VIEW f_score_view AS (
-- selecting distinct records for each customer based on unique customer_id and invoice_no (this shows unique transactions)
    WITH unique_invoices AS (
        SELECT DISTINCT
            customer_id, invoice_no 
        FROM orders
    ),
-- aggregating the number of transactions per customer
    frequency AS (
        SELECT
            customer_id,
            COUNT(invoice_no) AS frequency
        FROM unique_invoices
        GROUP BY customer_id
-- grouping the number of transactions per customer into 4 quartiles
    ),
    frequency_quartile AS (
        SELECT
            customer_id,
            frequency,
            NTILE(4) OVER(ORDER BY frequency) AS f_score
        FROM frequency
    )
SELECT *
FROM frequency_quartile
ORDER BY customer_id
)

 * postgresql://postgres:***@localhost:5432/rfm_nest_and_nook
Done.


[]

In [95]:
%%sql
-- previewing the f_score_view
SELECT *
FROM f_score_view
LIMIT 5

 * postgresql://postgres:***@localhost:5432/rfm_nest_and_nook
5 rows affected.


[(12346, 2, 2), (12347, 7, 4), (12348, 4, 3), (12349, 1, 1), (12350, 1, 1)]

## **Monetary**
- This is the total amount of money a specific customer has spent in purchasing a product. It is calculated by aggregating amount spent by each customer.

In [97]:
%%sql
-- CREATING A VIEW TO STORE EACH CUSTOMER AND THEIR MONETARY SCORE
CREATE VIEW m_score_view AS (
-- aggregating each customer's spend
    WITH total_spend AS (
        SELECT
            customer_id,
            SUM(ABS(total_sales)) AS monetary
        FROM orders
        GROUP BY customer_id
    ),
-- grouping each customer's total spend into 4 quartiles
    monetary_quartile AS (
        SELECT
            customer_id,
            monetary,
            NTILE(4) OVER(ORDER BY monetary) AS m_score
        FROM total_spend
    )
SELECT *
FROM monetary_quartile
ORDER BY customer_id
)

 * postgresql://postgres:***@localhost:5432/rfm_nest_and_nook
Done.


[]

In [98]:
%%sql
SELECT * FROM m_score_view LIMIT 5

 * postgresql://postgres:***@localhost:5432/rfm_nest_and_nook
5 rows affected.


[(12346, Decimal('154367.2'), 4),
 (12347, Decimal('4310.00'), 4),
 (12348, Decimal('1797.24'), 4),
 (12349, Decimal('1757.55'), 4),
 (12350, Decimal('334.4'), 2)]

## **Joining the Recency, Frequency and Monetary Views**

In [100]:
%%sql
-- CREATING THE RFM_SCORE TABLE
CREATE VIEW rfm_score AS (
    WITH joined_views AS (
        SELECT *
        FROM r_score_view AS r
        INNER JOIN f_score_view AS f
            USING(customer_id)
        INNER JOIN m_score_view
            USING(customer_id)
    ),
    rfm_scoring AS (
        SELECT *,
            r_score::TEXT || f_score::TEXT || m_score::TEXT AS rfm_score
        FROM joined_views
    )
SELECT *
FROM rfm_scoring
)

 * postgresql://postgres:***@localhost:5432/rfm_nest_and_nook
(psycopg2.errors.DuplicateTable) relation "rfm_score" already exists

[SQL: -- CREATING THE RFM_SCORE TABLE
CREATE VIEW rfm_score AS (
    WITH joined_views AS (
        SELECT *
        FROM r_score_view AS r
        INNER JOIN f_score_view AS f
            USING(customer_id)
        INNER JOIN m_score_view
            USING(customer_id)
    ),
    rfm_scoring AS (
        SELECT *,
            r_score::TEXT || f_score::TEXT || m_score::TEXT AS rfm_score
        FROM joined_views
    )
SELECT *
FROM rfm_scoring
)]
(Background on this error at: https://sqlalche.me/e/20/f405)


In [101]:
%%sql
SELECT * FROM rfm_score LIMIT 5

 * postgresql://postgres:***@localhost:5432/rfm_nest_and_nook
5 rows affected.


[(12346, 325, 1, 2, 2, Decimal('154367.2'), 4, '124'),
 (12347, 2, 4, 7, 4, Decimal('4310.00'), 4, '444'),
 (12348, 75, 2, 4, 3, Decimal('1797.24'), 4, '234'),
 (12349, 18, 3, 1, 1, Decimal('1757.55'), 4, '314'),
 (12350, 310, 1, 1, 1, Decimal('334.4'), 2, '112')]

## Creating Customer Segments Based on RFM Scores

---

#### 1. **Best Customers**: `(R=4, F=4, M=4)`
Customers who score high in all dimensions: high recency, high frequency, and high monetary value. Thesethe your top-performing customers.

---

#### 2. **Loyal Customers**: `(R=3-4, F=3-4, M=2-4)`
Customers with moderate recency and frequency but a decent spending pattern. These are regular customers but not as valuable as the best customers.

---

#### 3. **At Risk**: `(R=1-2, F=3-4, M=3-4)`
Customers who used to buy frequently or spent a lot but haven't been active recently.

---

#### 4. **Low Value**: `(R=1-2, F=1-2, M=1-2)`
Customers with low scores across all dimensions: not recent, infrequent, and low spending. These are the least engaged or least valuable customers.


In [104]:
%%sql
CREATE VIEW rfm_segment AS (
WITH customer_segment AS
    (SELECT *,
        CASE
        -- Best Customers: High in all R, F, M dimensions
        WHEN r_score = 4 AND f_score = 4 AND m_score = 4 THEN 'Best Customers'
        
        -- Loyal Customers: Moderate recency, frequency, and spending
        WHEN r_score BETWEEN 3 AND 4 AND f_score BETWEEN 3 AND 4 AND m_score BETWEEN 2 AND 4 THEN 'Loyal Customers'
        
        -- At Risk: Customers who spent a lot or bought frequently, but not recent anymore
        WHEN r_score BETWEEN 1 AND 2 AND f_score BETWEEN 3 AND 4 AND m_score BETWEEN 3 AND 4 THEN 'At Risk'
        
        -- Low Value: Low in recency, frequency, and monetary value
        WHEN r_score BETWEEN 1 AND 2 AND f_score BETWEEN 1 AND 2 AND m_score BETWEEN 1 AND 2 THEN 'Low Value'

        -- Catch-All: Ensure all remaining combinations are captured here
        ELSE 'Low Value'
    END AS customer_segment
FROM rfm_score
)
SELECT *
FROM customer_segment
)

 * postgresql://postgres:***@localhost:5432/rfm_nest_and_nook
Done.


[]

In [105]:
%%sql
-- JOINING THE RFM TABLE WITH CUSTOMER TABLE FOR FEATURE ANALYSIS
CREATE VIEW rfm_table AS (
    SELECT *
    FROM rfm_segment
    LEFT JOIN customer
        USING(customer_id)
)

 * postgresql://postgres:***@localhost:5432/rfm_nest_and_nook
Done.


[]

In [106]:
%%sql
SELECT *
FROM rfm_table
LIMIT 10

 * postgresql://postgres:***@localhost:5432/rfm_nest_and_nook
10 rows affected.


[(12346, 325, 1, 2, 2, Decimal('154367.2'), 4, '124', 'Low Value', 'United Kingdom', 'Female', 55),
 (12347, 2, 4, 7, 4, Decimal('4310.00'), 4, '444', 'Best Customers', 'Iceland', 'Male', 27),
 (12348, 75, 2, 4, 3, Decimal('1797.24'), 4, '234', 'At Risk', 'Finland', 'Female', 32),
 (12349, 18, 3, 1, 1, Decimal('1757.55'), 4, '314', 'Low Value', 'Italy', 'Female', 51),
 (12350, 310, 1, 1, 1, Decimal('334.4'), 2, '112', 'Low Value', 'Norway', 'Female', 37),
 (12352, 36, 3, 11, 4, Decimal('3466.67'), 4, '344', 'Loyal Customers', 'Norway', 'Male', 67),
 (12353, 204, 1, 1, 2, Decimal('89.0'), 1, '121', 'Low Value', 'Bahrain', 'Male', 48),
 (12354, 232, 1, 1, 1, Decimal('1079.40'), 3, '113', 'Low Value', 'Spain', 'Female', 50),
 (12355, 214, 1, 1, 1, Decimal('459.4'), 2, '112', 'Low Value', 'Bahrain', 'Female', 31),
 (12356, 22, 3, 3, 3, Decimal('2811.43'), 4, '334', 'Loyal Customers', 'Portugal', 'Male', 51)]

## Analysis

In [107]:
%%sql
WITH segment_proportion AS (
    SELECT customer_segment, COUNT(*) AS no_of_customers
    FROM rfm_table
    GROUP BY customer_segment
    ORDER BY no_of_customers DESC
)
SELECT customer_segment,
        no_of_customers,
        ROUND((no_of_customers/(SELECT SUM(no_of_customers) FROM segment_proportion)), 2)||'%' AS proportion
FROM segment_proportion

 * postgresql://postgres:***@localhost:5432/rfm_nest_and_nook
4 rows affected.


[('Low Value', 2400, '0.55%'),
 ('Loyal Customers', 1019, '0.23%'),
 ('Best Customers', 480, '0.11%'),
 ('At Risk', 473, '0.11%')]

In [112]:
%%sql
SELECT 
    customer_segment, 
    ROUND(AVG(recency)) AS avg_recency, 
    ROUND(AVG(frequency), 1) AS avg_frequency, 
    ROUND(AVG(monetary)) AS avg_spend,
    ROUND(AVG(age)) AS avg_age
FROM rfm_table
GROUP BY customer_segment
ORDER BY avg_spend DESC, avg_frequency DESC, avg_recency DESC;

 * postgresql://postgres:***@localhost:5432/rfm_nest_and_nook
4 rows affected.


[('Best Customers', Decimal('6'), Decimal('19.0'), Decimal('9453'), Decimal('47')),
 ('Loyal Customers', Decimal('21'), Decimal('6.4'), Decimal('2610'), Decimal('45')),
 ('At Risk', Decimal('107'), Decimal('5.6'), Decimal('2289'), Decimal('44')),
 ('Low Value', Decimal('136'), Decimal('1.6'), Decimal('518'), Decimal('46'))]

In [132]:
%%sql
SELECT 
    customer_segment, 
    (SELECT country FROM rfm_table GROUP BY country ORDER BY count(*) DESC LIMIT 1),
    (SELECT gender FROM rfm_table GROUP BY gender ORDER BY count(*) DESC LIMIT 1)    
FROM rfm_table
GROUP BY customer_segment

 * postgresql://postgres:***@localhost:5432/rfm_nest_and_nook
4 rows affected.


[('At Risk', 'United Kingdom', 'Female'),
 ('Best Customers', 'United Kingdom', 'Female'),
 ('Low Value', 'United Kingdom', 'Female'),
 ('Loyal Customers', 'United Kingdom', 'Female')]

## Customer Profile

In [153]:
%%sql
-- JOINING THE 2 TABLES
WITH proportion AS (
WITH segment_proportion AS (
    SELECT customer_segment, COUNT(*) AS no_of_customers
    FROM rfm_table
    GROUP BY customer_segment
    ORDER BY no_of_customers DESC
)
SELECT customer_segment,
        ROUND((no_of_customers/(SELECT SUM(no_of_customers) FROM segment_proportion)), 2)||'%' AS proportion
FROM segment_proportion
),
rfm AS (
    SELECT *
    FROM (
        SELECT 
            customer_segment, 
            ROUND(AVG(recency)) AS avg_recency, 
            ROUND(AVG(frequency), 1) AS avg_frequency, 
            ROUND(AVG(monetary)) AS avg_spend,
            ROUND(AVG(age)) AS avg_age
        FROM rfm_table
        GROUP BY customer_segment
        ORDER BY avg_spend DESC, avg_frequency DESC, avg_recency DESC)
),
feature As (
    SELECT 
        customer_segment, 
        (SELECT country FROM rfm_table GROUP BY country ORDER BY count(*) DESC LIMIT 1) AS most_commmon_country,
        (SELECT gender FROM rfm_table GROUP BY gender ORDER BY count(*) DESC LIMIT 1) AS most_commmon_gender  
    FROM rfm_table
    GROUP BY customer_segment
)
SELECT *
FROM proportion
INNER JOIN rfm
    USING(customer_segment)
INNER JOIN feature
    USING(customer_segment)
ORDER BY avg_spend DESC

 * postgresql://postgres:***@localhost:5432/rfm_nest_and_nook
4 rows affected.


[('Best Customers', '0.11%', Decimal('6'), Decimal('19.0'), Decimal('9453'), Decimal('47'), 'United Kingdom', 'Female'),
 ('Loyal Customers', '0.23%', Decimal('21'), Decimal('6.4'), Decimal('2610'), Decimal('45'), 'United Kingdom', 'Female'),
 ('At Risk', '0.11%', Decimal('107'), Decimal('5.6'), Decimal('2289'), Decimal('44'), 'United Kingdom', 'Female'),
 ('Low Value', '0.55%', Decimal('136'), Decimal('1.6'), Decimal('518'), Decimal('46'), 'United Kingdom', 'Female')]

## INSIGHTS FROM CUSTOMER SEGMENTATION

A total of 4372 customers were analyzed, and their distribution is as follows:

- 11% were categorized as Best Customers.
- 23% were classified as Loyal Customers.
- 11% were identified as At Risk customers.
- 55% fell into the Low Value customer category.

### Demography
- The most common country for all customer segments is the United Kingdom.
- The most common gender across all segments is Female.
- The average age of customers across segments ranges from 44 to 47 years old.
- Loyal Customers have a higher average frequency of purchases (6.4) compared to At Risk (5.6) and Low Value (1.6) customers, but lower than Best Customers (19.0).


## RECOMMENDATIONS

- **Retain Best Customers** by offering personalized loyalty programs and exclusive promotions as well as engage them with targeted communication and early access to products.

- **Re-engage At Risk Customers** by using win-back campaigns with special offers and addressing reasons for disengagement through feedback analysis.

- **Improve Low Value Customer Engagement** by creating offers to increase purchase frequency and incentives for repeat purchases or referrals.
  
- **Enhance Loyal Customer Experience** offering personalized product recommendations and exclusive benefits.

- **Target UK and Female Customers** and tailoring content and promotions to their preferences.

- Develop campaigns that align with the lifestyle and needs of people in the **age bracket of 44–47 years.**